### PR TITLE
[NOT-YET] Update documentation for butterfly supervisor

### PR DESCRIPTION
**Pull request by adamhjk**
* Updates documentation to reflect changes caused by butterfly

Don't merge this yet - it will ship, and we don't want it to.

Signed-off-by: Adam Jacob <adam@chef.io>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: www/source/docs/internals-supervisor.html.md [hunk 1]
```diff
@@ -16,52 +16,88 @@ Supervisor rings can be very large, comprising potentially thousands of supervis
...
 
-## Protocols
+## Butterfly
+
+Habitat uses a gossip protocol named "Butterfly". It is a variant of
+[SWIM](http://prakhar.me/articles/swim) for membership and failure detection
+(over UDP), and a ZeroMQ based variant of
+[Newscast](http://www.cs.unibo.it/bison/publications/ap2pc03.pdf) for
+gossip. This protocol provides failure detection, service
+discovery, and leader election to the habitat supervisor.
```
Comment (metadave) [inline www/source/docs/internals-supervisor.html.md]: habitat -> Habitat

File: www/source/docs/internals-supervisor.html.md [hunk 2]
```diff
@@ -16,52 +16,88 @@ Supervisor rings can be very large, comprising potentially thousands of supervis
...
 
-However, if the peer turns out to not be reachable from any of the randomly selected three peers, the peer that times out first generates a "suspect" rumor to the whole ring, at which the entire ring attempts to send that rumor to the peer marked suspect. If the suspect peer receives this rumor and is capable of responding, it will do so by gossipping an "alive" rumor and incrementing its incarnation version in the membership list. This incarnation trumps any "suspect" rumors.
+### Gossip
 
-If the "suspect" rumor also times out, then the peer is marked "confirmed" to indicate that it is truly dead, the confirmation rumor is gossipped around the membership list, and all members remove the confirmed-dead member from their list. They will never communicate with the confirmed-dead member again -- unless that member recovers, and communicates with them.
+Butterfly uses ZeroMQ to disseminate rumors throughout the network. It's flow:
```
Comment (juliandunn) [inline www/source/docs/internals-supervisor.html.md]: `It's` -> `Its`

File: www/source/docs/run-packages-upload-files.html.md [hunk 3]
```diff
@@ -3,44 +3,42 @@ title: Upload files to a service group
 ---
 
 # Upload files to a service group
-In addition to [configuration updates](/docs/run-packages-apply-config-updates), you can upload files to a service group. The file upload process is very similar to applying configuration updates; however, there is one important difference: You must use [service group encryption](/docs/run-packages-security#service-group-encryption) to upload files to a service group. This requires creating service keys and user keys. File uploads cannot be in the clear.
+In addition to [configuration updates](/docs/run-packages-apply-config-updates), you can upload files to a service group. Keep these small - we recommend 4k or less per file, and keep them to a minimum.
```
Comment (juliandunn) [inline www/source/docs/run-packages-upload-files.html.md]: "Keep them to a minimum" is redundant unless you're trying to say something that "keep these small" isn't already saying.

**Comment by reset** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by reset** [comment]
All feedback encorporated, thanks everyone!

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/184033817).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/184033817).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**